app/models/similarity/unit_similarity_module.rb

```diff
@@ -222,4 +235,51 @@ def add_done_files_for_plagiarism_check_of(task_definition, tmp_path, to_check,
 
     self
   end
+
+  # JPLAG Function - extracts "done" files for each task and packages them into a directory for JPLAG to run on
+  def run_jplag_on_done_files(task_definition, tasks_dir, tasks_with_files, unit_code)
+    type_data = task_definition.moss_language.split
+    return if type_data.nil? || (type_data.length != 2) || (type_data[0] != 'moss')
+
+    # get each code file for each task
+    task_definition.upload_requirements.each_with_index do |upreq, idx|
+      # only check code files marked for similarity checks
+      next unless upreq['type'] == 'code' && upreq['tii_check']
+
+      pattern = task_definition.glob_for_upload_requirement(idx)
+
+      tasks_with_files.each do |t|
+        t.extract_file_from_done(tasks_dir, pattern, ->(_task, to_path, name) { File.join(to_path.to_s, t.student.username.to_s, name.to_s) })
+      end
+      puts "Starting JPLAG container to run on #{tasks_dir}"
+      tasks_dir_split = tasks_dir.to_s.split('/workspace/doubtfire-api')[1]
+
+      # Check if the directory exists and create it if it doesn't
+      results_dir = "/jplag/results/#{unit_code}"
+      `sudo docker exec jplag sh -c 'if [ ! -d "#{results_dir}" ]; then mkdir -p "#{results_dir}"; fi'`
```

Comment:
sudo?

Reply:
Should this be outside the upload requirements loop? What happens if you have multiple files being checked in the same task definition?

Reply:
Sudo is needed to make the `docker exec` commands work. Not entirely sure why 🙃.
Have moved the dir checks outside the loop and still works as expected.

Reply:
`docker exec` doesn't require `sudo` in production since it'll run as the root user, but the permission errors in the devcontainer are due to it using the `vscode` user

You can fix this by [setting permissions on `/var/run/docker.sock`](https://github.com/doubtfire-lms/doubtfire-deploy/pull/29/commits/a3af75f200fdb4be9ce5afc621c906606bbbeb73#diff-79495a5ff44c04cd52db3561d48e92865b1b47e8738ddc27a489224de4914daaR10) in https://github.com/doubtfire-lms/doubtfire-deploy